modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/RepositoryStatsCollector.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.repositories.gcs;
+
+import org.elasticsearch.common.CheckedSupplier;
+import org.elasticsearch.common.blobstore.BlobStoreActionStats;
+import org.elasticsearch.common.blobstore.OperationPurpose;
+import org.elasticsearch.core.CheckedRunnable;
+import org.elasticsearch.repositories.RepositoriesMetrics;
+
+import java.io.IOException;
+import java.util.EnumMap;
+import java.util.Map;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class RepositoryStatsCollector {
+
+    /**
+     * Track billable requests only, required for REST repository metering API
+     */
+    private final EnumMap<StorageOperation, LongAdder> restMetering;
+
+    /**
+     * Telemetry (APM)
+     */
+    private final RepositoriesMetrics telemetry;
+    private final EnumMap<OperationPurpose, EnumMap<StorageOperation, Map<String, Object>>> telemetryAttributes;
+
+    RepositoryStatsCollector(String repoName, RepositoriesMetrics repositoriesMetrics) {
+        this.telemetry = repositoriesMetrics;
+        this.restMetering = new EnumMap<>(StorageOperation.class);
+        for (var op : StorageOperation.values()) {
+            restMetering.put(op, new LongAdder());
+        }
+        this.telemetryAttributes = new EnumMap<>(OperationPurpose.class);
+        for (var purpose : OperationPurpose.values()) {
+            var purposeMap = new EnumMap<StorageOperation, Map<String, Object>>(StorageOperation.class);
+            telemetryAttributes.put(purpose, purposeMap);
+            for (var operation : StorageOperation.values()) {
+                var attrMap = RepositoriesMetrics.createAttributesMap(GoogleCloudStorageRepository.TYPE, repoName, purpose, operation.key);
+            }
+        }
+    }
+
+    /**
+     * Continue collecting metrics with given OperationStats. Useful for readers and writers.
+     */
+    public <T> T continueAndCollect(OperationStats stats, CheckedSupplier<T, IOException> blobFn) throws IOException {
+        OperationStats.set(stats);
+        return blobFn.get();
+    }
+
+    /**
+     * Continue collecting metrics with given OperationStats. Useful for readers and writers.
+     */
+    public void continueAndCollect(OperationStats stats, CheckedRunnable<IOException> runnable) throws IOException {
+        OperationStats.set(stats);
```

Comment:
I wonder if we should be more assertive about the current state of `OperationStats`. In this case it's probably safe to clear it after each action right? then we could assert that it was empty each time we went to start a new action, to make it fail noisily if they started being interleaved somehow?